test/envtest/gatewayclass_envtest_test.go

```diff
@@ -0,0 +1,302 @@
+)
+
+func init() {
+	if err := gatewayv1beta1.Install(scheme.Scheme); err != nil {
+		panic(err)
+	}
+}
+
+func TestGatewayClassReconciliation(t *testing.T) {
+	t.Parallel()
+
+	const (
+		// unsupportedControllerName is the name of the controller used for those
+		// gateways that are not supported by an actual controller (i.e., they won't be scheduled).
+		unsupportedControllerName gatewayv1beta1.GatewayController = "example.com/unsupported-gateway-controller"
+	)
+
+	cfg := Setup(t, scheme.Scheme)
+
+	gatewayClient, err := gatewayclient.NewForConfig(cfg)
+	require.NoError(t, err)
+
+	client, err := ctrlclient.New(cfg, ctrlclient.Options{
+		Scheme: scheme.Scheme,
+	})
+	require.NoError(t, err)
+
+	testcases := []struct {
+		Name         string
+		GatewayClass gatewayv1beta1.GatewayClass
+		Gateway      gatewayv1beta1.Gateway
+		Test         func(context.Context, *testing.T, gatewayclientv1beta1.GatewayInterface, gatewayv1beta1.GatewayClass, gatewayv1beta1.Gateway)
+	}{
+		{
+			Name: "unsupported gateway class",
+			GatewayClass: gatewayv1beta1.GatewayClass{
+				Spec: gatewayv1beta1.GatewayClassSpec{
+					ControllerName: unsupportedControllerName,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "unsupported-gateway-class",
+				},
+			},
+			Gateway: gatewayv1beta1.Gateway{
+				Spec: gatewayv1beta1.GatewaySpec{
+					GatewayClassName: gatewayv1beta1.ObjectName("unsupported-gateway-class"),
+					Listeners: builder.NewListener("http").
+						HTTP().
+						WithPort(80).
+						WithAllowedRoutes(builder.NewAllowedRoutesFromAllNamespaces()).
+						IntoSlice(),
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: uuid.NewString(),
+				},
+			},
+			Test: func(ctx context.Context, t *testing.T, gwClient gatewayclientv1beta1.GatewayInterface, gwc gatewayv1beta1.GatewayClass, gw gatewayv1beta1.Gateway) {
+				t.Helper()
+
+				t.Logf("deploying gateway class %s", gwc.Name)
+				require.NoError(t, client.Create(ctx, &gwc))
+				t.Cleanup(func() { _ = client.Delete(ctx, &gwc) })
+
+				t.Logf("verifying that the unsupported Gateway %s does not get Accepted or Programmed by the controller", gw.Name)
+				for i := 0; i < 100; i++ {
+					gateway, err := gwClient.Get(ctx, gw.Name, metav1.GetOptions{})
+					require.NoError(t, err)
+					require.Len(t, gateway.Status.Conditions, 2)
+					if lo.ContainsBy(gateway.Status.Conditions, func(c metav1.Condition) bool {
+						return c.Type == "Programmed" && c.Status != metav1.ConditionUnknown
+					}) {
+						t.Fatalf("expected not to find a Programmed condition with Reason different than Unknown, got %#v", gateway.Status.Conditions)
+					}
+					if lo.ContainsBy(gateway.Status.Conditions, func(c metav1.Condition) bool {
+						return c.Type == "Accepted" && c.Status != metav1.ConditionUnknown
+					}) {
+						t.Fatalf("expected not to find a Accepted condition with Reason different than Unknown, got %#v", gateway.Status.Conditions)
+					}
+				}
```

Comment:
This fragment is duplicated in a couple of test cases, maybe worth extracting a helper for it?

Reply:
Refactored this out into `test/helpers/conditions/` package and used it in this PR.